staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -218,16 +217,21 @@ func (s *specAggregator) AddUpdateAPIService(apiService *v1.APIService, handler
 	s.mutex.Lock()
 	defer s.mutex.Unlock()
 
-	_, exists := s.specsByAPIServiceName[apiService.Name]
+	existingSpec, exists := s.specsByAPIServiceName[apiService.Name]
 	if !exists {
-		s.specsByAPIServiceName[apiService.Name] = &openAPISpecInfo{
+		specInfo := &openAPISpecInfo{
 			apiService: *apiService,
-			downloader: decorateError(apiService.Name, NewCacheableDownloader(s.downloader, handler)),
+			downloader: NewCacheableDownloader(apiService.Name, s.downloader, handler),
 		}
+		specInfo.spec.Store(cached.Result[*spec.Swagger]{Err: fmt.Errorf("spec for apiservice %s is not yet available", apiService.Name)})
+		s.specsByAPIServiceName[apiService.Name] = specInfo
 		s.openAPIVersionedService.UpdateSpecLazy(s.buildMergeSpecLocked())
```

Comment:
I guess buildMergeSpecLocked already called Get(), which meant a slow request / timeout of one service would keep the specAggregator#UpdateAPIServiceSpec loop from processing other API services. Is that new in 1.28 or was it existing?

Reply:
specAggregator#UpdateAPIServiceSpec blocking for each apiservice is pre-existing. The controller is single threaded so the UpdateAPIServiceSpec loop processes one apiservice at a time in a blocking manner. builtMergeSpecLocked calling Get() is a new addition, it is only called for local apiservices and from observation local apiservices are very unlikely to block and cause problems.

Reply:
> specAggregator#UpdateAPIServiceSpec blocking for each apiservice is pre-existing

does that mean this PR does not resolve https://github.com/kubernetes/kubernetes/issues/120739#issuecomment-1725875180

> The issue occurs only after v1.28 may be because https://github.com/kubernetes/kubernetes/pull/118212 changed c.openAPIAggregationManager.AddUpdateAPIService to download the latest spec synchronously?

Reply:
when an APIService is added, only AddUpdateAPIService is called. There was a bug where AddUpdateAPIService called UpdateAPIServiceSpec that was fixed by this commit targeting #120739: https://github.com/kubernetes/kubernetes/pull/120814/commits/61603d36d738dddf3a7f244e0f8630f9ebeeffaf.

UpdateAPIServiceSpec is called by the openapi controller and is not in the path for readiness, and we are essentially leaving the behavior of UpdateAPIServiceSpec unchanged from before the regression.

Reply:
sorry, I think I copied the wrong bit from your comment

> builtMergeSpecLocked calling Get() is a new addition

Aren't we still doing that in the current state of this PR?

> it is only called for local apiservices and from observation local apiservices are very unlikely to block and cause problems.

Where is it limited to local apiservices? This function appears to only be handling *remote* API services (it short-circuits if apiService.Spec.Service == nil)

Reply:
> Aren't we still doing that in the current state of this PR?

Yes, `buildMergeSpecLocked` calls `Get()` on local apiservices. The problem described in the issue is for remote apiservices which leads to your next comment

> Where is it limited to local apiservices? This function appears to only be handling remote API services (it short-circuits if apiService.Spec.Service == nil)

`buildMergeSpecLocked()` calling `Get()` is limited to local apiservices. The `UpdateAPIServiceSpec()` function only affects remote apiservices (as seen with the short circuit). `UpdateAPIServiceSpec()` is not called within the critical path of readiness (only AddUpdateAPIService).

For readiness: 
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L330-L337
The apiservice controller syncs all existing apiservices before closing the ready channel and calls AddAPIService for each existing apiservice:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiservice_controller.go#L112

which goes through
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L478

and then reaches openapi's AddUpdateAPIService.

Reply:
> `buildMergeSpecLocked()` calling `Get()` is limited to local apiservices.

It took me a while to understand why that is true...
1. we populate a cache entry for this remote service here containing an error that the spec wasn't synced yet
```go
		specInfo.spec.Store(cached.Result[*spec.Swagger]{Err: fmt.Errorf("spec for apiservice %s is not yet available", apiService.Name)})
		s.specsByAPIServiceName[apiService.Name] = specInfo
```
2. buildMergeSpecLocked accumulates cached specs:
```
	for i, apiService := range apiServices {
		caches[i] = &(s.specsByAPIServiceName[apiService.Name].spec)
	}
```
3. buildMergeSpecLocked merges calling .Get on each one:
```go
	return cached.MergeList(func(results []cached.Result[*spec.Swagger]) (*spec.Swagger, string, error) {
		var merged *spec.Swagger
		etags := make([]string, 0, len(results))
		for _, specInfo := range results {
			result, etag, err := specInfo.Get()
```
4. the result we just included will just return an error:
```go
func (r Result[T]) Get() (T, string, error) {
	return r.Value, r.Etag, r.Err
}
```
5. which buildMergeSpecLocked will log and continue:
```go
			if err != nil {
				// APIService name and err message will be included in
				// the error message as part of decorateError
				klog.Warning(err)
				continue
			}
```

I didn't realize we were adding two different implementations of Get() into `specsByAPIServiceName`, and that the one for remote services had a non-blocking Get() implementation

Reply:
you folks are confusing to read, `buildMergeSpecLocked` almost does nothing, the closure that it builds does most of the work ... asynchronously. But yes, otherwise

> I didn't realize we were adding two different implementations of Get() into specsByAPIServiceName, and that the one for remote services had a non-blocking Get() implementation

That's correct.